google/cloud/storage/internal/invocation_id_generator.cc

```diff
@@ -0,0 +1,67 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/invocation_id_generator.h"
+#include "absl/strings/str_format.h"
+#include <algorithm>
+#include <array>
+
+namespace google {
+namespace cloud {
+namespace storage {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+// Initialize the random bit source with some minimal entropy.
+InvocationIdGenerator::InvocationIdGenerator()
+    : generator_(std::random_device{}()) {}
+
+std::string InvocationIdGenerator::MakeInvocationId() {
+  // A retry id is supposed to be a UUID V4 string. We assume you have read the
+  // wikipedia page for the details:
+  //     https://en.wikipedia.org/wiki/Universally_unique_identifier
+  auto constexpr kVersionOctet = 6;
+  auto constexpr kVariantOctet = 8;
+  auto constexpr kVersion = static_cast<std::uint8_t>(4 << 4);
+  auto constexpr kVersionMask = static_cast<std::uint8_t>(0b0000'1111);
+  auto constexpr kVariant = static_cast<std::uint8_t>(1 << 5);
+  auto constexpr kVariantMask = static_cast<std::uint8_t>(0b0001'1111);
+
+  auto o = [this]() {
+    auto constexpr kIdBitCount = 128;
+    auto constexpr kArraySize = kIdBitCount / 8;
+    std::array<std::uint8_t, kArraySize> buf;
+    std::uniform_int_distribution<std::uint8_t> d(0, 255);
```

Comment:
I'm seeing a static assertion failure with a platform of interest.
```
    static_assert(__libcpp_random_is_valid_inttype<_IntType>::value, "IntType must be an integer type larger than char");
```

Reply:
For what it's worth, cppreference says ...
_"IntType - The result type generated by the generator. The effect is undefined if this is not one of short, int, long, long long, unsigned short, unsigned int, unsigned long, or unsigned long long."_

Reply:
Blegh.  Okay, we can use `int` and cast to the desired result.  The output should be in range.  I can send a PR, but I need to dig myself from all the upgrade PRs that are breaking things right now.

Reply:
See #9509. Let's see what the builds say.